### PR TITLE
Created an automated repo environment

### PR DESCRIPTION
**Pull request by cloudnull**
This adds the ability for the user to stand up a repo for all of the python packages
that may be needed for a given environment. This will allow a user to maintain their
own repo of packages for production purposes in CDC where public internet may not be
available or for development purposes.
- The repo created by these plays will create a python wheel archive and a git server powered by NGINX enables support on port 8181 via haproxy for testing, and moves the wheel builder scripts into the `pkg_repo` roles. 
- The plays also have support for multi node environments which will provide for scalability of the internal repo. This is provided by lsync on the multiple nodes. 
- The repo storage is mounted in /openstack on the host.
- The cloudserver-aio script has been updated with the new container type. 

Related Issue: https://github.com/rcbops/u-suk-dev/issues/48
Related Issue: https://github.com/rcbops/ansible-lxc-rpc/issues/527

### CODE REVIEW

**Comment by hughsaunders** [comment]
```
06:05:04 PLAY [nova_compute[0]] ******************************************************** 
06:05:04 
06:05:04 TASK: [Distribute authorized keys for cluster consumption] ******************** 
06:05:04 failed: [node37] => (item={'src': '/var/lib/nova/.ssh/authorized_keys', 'name': 'authorized_keys'}) => {"err": "Memcache content not created", "failed": true, "item": {"name": "authorized_keys", "src": "/var/lib/nova/.ssh/authorized_keys"}, "rc": 1}
06:05:04 msg: The content you attempted to place within memcached was not created. If you are load balancing memcached, attempt to connect to a single node. Returned a value of unstored keys [ ['authorized_keys.0'] ].
06:05:04 
06:05:04 FATAL: all hosts have already failed -- aborting
```

I think this gate failure is genuine, it happens consistently with this PR, but I haven't been able to reproduce the failure with master. See http://jenkins.propter.net/job/dev-sat6-gate/344/console for example. 

```
$ for i in {342..1}; do curl http://jenkins.propter.net/job/dev-sat6-gate/$i/consoleText >/tmp/gatelog 2>&1; mcf=$(grep -q "Memcache content not created" </tmp/gatelog && echo -n "Memcache Fail "); prid=$(grep -o "pull request #[0-9]*" </tmp/gatelog |tail -n1|tr -d '\n'); echo "Job:$i $prid $mcf";  done
Job:342 pull request #543 Memcache Fail
Job:341 pull request #543 Memcache Fail
Job:340 pull request #543 Memcache Fail
Job:339 pull request #543 Memcache Fail
Job:338 pull request #544
Job:337 pull request #543 Memcache Fail
Job:336 pull request #543 Memcache Fail
Job:335 pull request #543 Memcache Fail
Job:334 pull request #528
Job:333 pull request #528
Job:332 pull request #543
Job:331 pull request #542
Job:330 pull request #540
Job:329 pull request #528
Job:328 pull request #540
Job:327 pull request #540
Job:326 pull request #523
Job:325 pull request #539
Job:324 pull request #538
Job:323 pull request #537
Job:322 pull request #536
Job:321 pull request #535
Job:320 pull request #534
```

**Comment by nrb** [comment]
I'm running into this doing this manually, too. I'm still investigating, though.

**Comment by nrb** [comment]
Ok, in my case, I had the IP address wrong in the `memcached:vars:memcached_listen` variable. I have not tried this with a full RPC setup yet, though.

**Comment by nrb** [comment]
Also, I have a request - can this gist be included somewhere? https://gist.github.com/cloudnull/fbcd4e371216af2fef31#deploying-on-metal-outside-rpc

File: etc/rpc_deploy/rpc_environment.yml [hunk 1]
```diff
@@ -335,3 +343,9 @@ physical_skel:
   swift-proxy_hosts:
     belongs_to:
     - hosts
+  repo_hosts:
+    belongs_to:
+    - hosts
+  repo_containers:
+    belongs_to:
+    - all_containers
```
Comment (nrb) [inline etc/rpc_deploy/rpc_environment.yml:351]: This line means that when the `host-setup.yml` playbook runs and calls [containers_common.yml](https://github.com/rcbops/ansible-lxc-rpc/blob/master/rpc_deployment/playbooks/setup/containers-common.yml#L16)  the repo containers will be included in the `common` role, which installs a `pip.conf` file ([`get_pip.yml`](https://github.com/rcbops/ansible-lxc-rpc/blob/master/rpc_deployment/roles/common/tasks/get_pip.yml#L41)).

Given what the repo containers do, we'll probably want a way to opt them out of this behavior.

**Comment by odyssey4me** [comment]
Does this mean that a deployment can be done with all hosts barring the repo container (and its hosting servers) can be deployed without internet access? If so, this could be a different sort of win where deployments could be done with the repo 'frozen-in-time' for the deployment. The repo could then be updated as an when the customer is happy to do so according to their change management policies.

**Comment by nrb** [comment]
@cloudnull I've provided a pull request against your branch that allowed me to get past my pip errors (once I removed the originally dropped pip configs). It's at https://github.com/cloudnull/ansible-lxc-rpc/pull/3

**Comment by nrb** [comment]
@odyssey4me As I understand, yes, this would accommodate that scenario. These plays are what power our current repos, so this would give users their own, locally contained repo they could use.

**Comment by cloudnull** [comment]
@nrb I've updated the repo PR to incorporate your changes though in a slightly different way. The changes were done in commit https://github.com/cloudnull/ansible-lxc-rpc/commit/095083a49cd5e3489a5ebe9747f46b669beba870.

@odyssey4me Yes the intention is to allow a customer to host their own repos and update them whenever they choose to. The aim was to open up what was running our present repos as well as provide a method for a customer to run in their DC which may or may not have public internet access.

**Comment by nrb** [comment]
@cloudnull Cool, I'm still going through the rest of my install, but your method is cleaner.

**Comment by cloudnull** [comment]
@hughsaunders - looks like gating was doing a great job. this commit changed the workflow for the memcache install which which resulted a variable that was missing. This has been resolved in 8b3083b

**Comment by nrb** [comment]
Thanks for that update Kevin - I just hit this in my AIO, but I'll update my branches to reflect this.

**Comment by cloudnull** [comment]
@nrb  yea my bad, i should've communicated that issue sooner.

**Comment by cloudnull** [comment]
@hughsaunders @nrb @odyssey4me - i just squashed the commits.

**Comment by nrb** [comment]
I'm still seeing the pip config files being dropped, even with the lock down file. I'm still investigating. It appears to progress when I delete the config files and re-run `repo-install.yml`.

**Comment by nrb** [comment]
With the latest round of commits, I'm able to proceed past the `repo-install.yml` step. I'm moving on to make sure the OpenStack playbooks work, too.

**Comment by nrb** [comment]
I've been able to successfully deploy an AIO with the latest round of changes to this PR, so 👍 from me.

**Comment by cloudnull** [comment]
test this please

**Comment by cloudnull** [comment]
This errors here are due to issues that do not effect the execution of the code added/modified in this PR.

**Comment by cloudnull** [comment]
Recent failure was with the installation of OpenJDK.

``` bash
18:33:13 TASK: [logging_common | Install OpenJDK] ************************************** 
18:33:19 failed: [node17_elasticsearch_container-90cce5fd] => {"failed": true, "parsed": false}
```

**Comment by cloudnull** [comment]
moved https://review.openstack.org/#/c/139260/

**Pull request closed without merging**